### ISSUE TITLE
Inference fails for intersection of types with a `toString` property on a branch 

### ISSUE THREAD

**Issue by bogdanb** [microsoft/TypeScript#54345]
# Bug Report

### 🔎 Search Terms

toString 

### 🕗 Version & Regression Information

- This changed between versions 4.8.4 and 4.9.5

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgCoQM5gIzIN7JgD2AymFKAOYBcyIArgDaPIA+yA5FhSJRwNzIAvgFgAUKEixEKdFgBMyCAA9IIACYY0mHMnzIiABzDAiIOIwD8teiADWIIgHcQw-uLABPQ7J0BmZABebSxcADJ9IxMzC2tkWwdnVyF3MXEEMyxCHQAGWjldYIJiMh4aOiYWFOQAehrkAHkAaXTMsGzQ-J1cIuQAOgHILBy3Wvrm1pAsobB5LoUg-QG+mZHqusaWsQyp9pm-ebAA3uXV0Y34YEYMcSA)

### 💻 Code

```ts
interface Test1 { toString: null | 'string'; }
interface Test2 extends Test1  { optional?: unknown };
type Test3 = Test1 & { optional?: unknown };

const test0: Test1 = { toString: null }; // OK
const test1: Test1 = { ...test0 }; // OK
const test2: Test2 = { ...test0 }; // OK
const test3: Test3 = { ...test0 }; // fails
```

### 🙁 Actual behavior

TypeScript complains about the declaration of `test2`:

```
Type '{ toString: "string" | null; }' is not assignable to type 'Test3'.
  Types of property 'toString' are incompatible.
    Type '"string" | null' is not assignable to type '"string" & (() => string)'.
      Type 'null' is not assignable to type '"string" & (() => string)'.
        Type 'null' is not assignable to type '"string"'.
```

Presumably it gets confused at some point by `Object.prototype.toString`.

Note that all the other tests work correctly.

### 🙂 Expected behavior

The code should compile as given without errors.

**Comment by Andarist** [comment]
I'm actually not sure what is the changed behavior that you are reporting here. From what I can tell we get the same results in 4.8.4 and 4.9.5 for the given code snippet.

**Comment by bogdanb** [comment]
OK, that is super weird, I get the same results too. I swear I tried it several times when I reported this, and it acted differently between those two versions. No idea what went wrong when.

(Originally I started getting this error in code that used to work, in a commit where we updated the TS version. I still get the behavior change around that commit, but perhaps it’s caused by a different change, I’ll keep investigating and report here if I find anything relevant.)

That said, even if this isn’t a regression as I thought, it still seems like a bug to me. I could not find anything in TypeScript that would make the declaration of `test3` illegal but allow the other ones. The error suggests that TS thinks `Test3.toString` is supposed to be assignable to `"string" & (() => string)`, and there seems to be no good reason for that. AFAIK, types `Test2` and `Test3` should behave pretty much identically.

**Comment by bogdanb** [comment]
So I investigated some more, it seems there is a relevant difference. Running TSC on the code below behaves differently between versions, it works with 4.8.4 but shows an error on 4.9.5:

```
interface Test1 { toString: string | null; }
type Test2 = Test1 & { optional?: unknown };
interface Test3 { prop: Test2 }

const x: Test1 = {toString: null};
const y = { ...x };
const z : Test3 = {prop: y};
```

Playground links: [4.8.4](https://www.typescriptlang.org/play?ts=4.8.4#code/JYOwLgpgTgZghgYwgAgCoQM5gIzIN7JgD2AymFKAOYBcyWFIlyAPsiAK4A2nA3MgL4AoMAE8ADinRYATMgC8aTDmQAyfMiJiwwIiDicA-LXYgA1iCIB3EAJ6DQkWIklKAzOrFRNtKWFlDBBF0sZAAPHyVcBTxiMgYaNi5OfjsgkBCReXUAOlzQ20DgsGQAL2QIrHdoz29kERSgA) / [4.9.5](https://www.typescriptlang.org/play?ts=4.9.5#code/JYOwLgpgTgZghgYwgAgCoQM5gIzIN7JgD2AymFKAOYBcyWFIlyAPsiAK4A2nA3MgL4AoMAE8ADinRYATMgC8aTDmQAyfMiJiwwIiDicA-LXYgA1iCIB3EAJ6DQkWIklKAzOrFRNtKWFlDBBF0sZAAPHyVcBTxiMgYaNi5OfjsgkBCReXUAOlzQ20DgsGQAL2QIrHdoz29kERSgA)

If we instead declare `const y: Test2`, then it has an error with both TS versions. So there _is_ a difference in the type inference, but it’s not exactly a regression, it just exposes an existing bug in this particular situation, where we pass through an intermediary variable that is not explicitly typed.

(I still don’t understand why I got different results in the playground when I entered the issue. Maybe I mixed up different test code trying to create a minimal test case.)
